lib/pure/net.nim

```diff
@@ -191,6 +191,31 @@ proc isDisconnectionError*(flags: set[SocketFlag],
     SocketFlag.SafeDisconn in flags and
       lastError.int32 in {ECONNRESET, EPIPE, ENETRESET}
 
+proc checkCloseError*(ret: cint) =
+  ## Asserts that the return value of close() or closeSocket() syscall
+  ## does not indicate a programming error (such as invalid descriptor).
+  ## This must only be used when an error has already occurred and
+  ## you are performing a cleanup.
+  ## Otherwise, error handling must be performed as usual.
+  ##
+  ## This procedure must be called right after perfoming the syscall. Example:
+  ##
+  ## .. code-block:: nim
+  ##
+  ##  let ret = someSysCall()
+  ##  if ret != 0:
+  ##    let errcode = osLastError()
+  ##    checkCloseError sock.closeSocket()
+  ##    raise newException(OSError, osErrorMsg(errcode))
+
+  if ret != 0:
+    let errcode = osLastError()
+    when useWinVersion:
+      doAssert(errcode.int32 notin {WSANOTINITIALISED, WSAENOTSOCK,
+                                    WSAEINPROGRESS, WSAEINTR, WSAEWOULDBLOCK})
```

Comment:
This is what you asked to do in the first place, please read your first comment again. Yes, `doAssert` it raises an AssertionError and app code shouldn't normally catch it, as it signalises programming error. This procedure discards OS errors and raises programming errors.

Again, this proc is to be used when you are already preparing to throw an OS error. We are not trying to prevent OS errors here, this just replaces `discard` that was used in the initial PR. We could make the close() throw its error as OS error, but then the initial error would be hidden, and I think it's better to show the first error than the second. Alternatively, we could implement something that stacks multiple OS errors into a single exception.